### PR TITLE
Improve chinese module

### PR DESCRIPTION
**Pull request by merrickluo**
I've made some improvements to the Chinese module and has been using it for years, here is what I add

- added a `+childframe` feature flag to make `pyim` show candidates in a childframe (using posframe).
- added a `+rime` feature flag to make pyim use [librime](https://github.com/rime/librime) as input method engine (new package [liberime](https://github.com/merrickluo/liberime)).
- use `pyim-cregexp-build` to make chinese characters can be searched with pinyin, similar to #6356 but without adding new package.
- disable `evil-snipe` when input method is activated, it intercepts key event that can mess up the input method.

-----
- [x] I searched the issue tracker and this hasn't been PRed before.
- [x] My changes are not on [the do-not-PR list](https://doomemacs.org/d/do-not-pr) for this project.
- [x] My commits conform to [the git conventions](https://doomemacs.org/d/git-conventions).
- [x] Any relevant issues or PRs have been linked to.

### CODE REVIEW

File: modules/input/chinese/config.el [hunk 1]
```diff
@@ -6,7 +6,41 @@
...
+  :when (featurep! +rime)
+  :init
+  (setq liberime-auto-build t)
+  (setq liberime-user-data-dir (expand-file-name "rime" doom-cache-dir)))
+
+(use-package! pyim-liberime
+  :when (featurep! +rime)
+  :after liberime
+  :config
+  (setq pyim-default-scheme 'rime))
```
Comment (hlissner) [inline modules/input/chinese/config.el:43]: Please separate top-level `use-package` blocks with two blank lines.

File: modules/input/chinese/config.el [hunk 2]
```diff
@@ -6,7 +6,41 @@
...
+         (defadvice! +pinyin-orderless-regexp (orig-fn component)
+           :around 'orderless-regexp
+           (let ((result (funcall orig-fn component)))
+             (pyim-cregexp-build result))))
+        ((featurep! :completion ivy)
+         (setq ivy-re-builders-alist
+               '((t . pyim-cregexp-ivy))))))
+
+(use-package! posframe
+  :when (featurep! +childframe))
```
Comment (hlissner) [inline modules/input/chinese/config.el:31]: Lazy-load posframe. If pyim is expected to use it, use `:after pyim`.

File: modules/input/chinese/packages.el [hunk 3]
```diff
@@ -5,3 +5,7 @@
 (package! fcitx :pin "12dc2638ddd15c8f6cfaecb20e1f428ab2bb5624")
 (package! ace-pinyin :pin "47662c0b05775ba353464b44c0f1a037c85e746e")
 (package! pangu-spacing :pin "f92898949ba3bf991fd229416f3bbb54e9c6c223")
+(when (featurep! +rime)
+  (package! liberime :pin "8291e22cd0990a99cb2f88ca67a9065a157f39af"))
+(when (featurep! +childframe)
+  (package! posframe))
```
Comment (hlissner) [inline modules/input/chinese/packages.el:11]: Packages shouldn't be unpinned. Use `(package! posframe :pin "0d23bc5f7cfac00277d83ae7ba52c48685bcbc68")`

File: modules/input/chinese/config.el [hunk 4]
```diff
@@ -6,7 +6,41 @@
   (setq pyim-dcache-directory (concat doom-cache-dir "pyim/"))
   :config
   (setq pyim-page-tooltip t
-        default-input-method "pyim"))
+        default-input-method "pyim")
+
+  (defun +input-method-activate-p ()
+    current-input-method)
```
Comment (hlissner) [inline modules/input/chinese/config.el:12]: Move this inside the `(after! evil-escape ...)` block, as it isn't used anywhere else.

File: modules/input/chinese/config.el [hunk 5]
```diff
@@ -6,7 +6,41 @@
...
+
+  (after! posframe
+    (setq pyim-page-tooltip 'posframe))
+
+  ;; allow vertico/selectrum search with pinyin
+  (cond ((featurep! :completion vertico)
+         (defadvice! +pinyin-orderless-regexp (orig-fn component)
+           :around 'orderless-regexp
+           (let ((result (funcall orig-fn component)))
+             (pyim-cregexp-build result))))
```
Comment (hlissner) [inline modules/input/chinese/config.el:25]: 1. Use a `:filter-return` advice here instead.
2. Rename advice to `+pinyin-orderless-regexp-a` to conform to [naming conventions](https://docs.doomemacs.org/latest/#/developers/conventions/emacs-lisp/naming-conventions/doom-conventions).
3. Sharp quote function symbols (`#'orderless-regexp`).

```diff
  (cond ((featurep! :completion vertico)
-         (defadvice! +pinyin-orderless-regexp (orig-fn component)
-           :around 'orderless-regexp
-           (let ((result (funcall orig-fn component)))
-             (pyim-cregexp-build result))))
+         (defadvice! +pinyin-orderless-regexp-a (result)
+           :filter-return #'orderless-regexp
+           (pyim-cregexp-build result)))
         ((featurep! :completion ivy)
```

**Review by hlissner — CHANGES_REQUESTED** [review]

**Comment by hlissner** [comment]
Can this PR salvage anything from #6482? I'd like to merge them into one, if possible. 

cc @merrickluo

File: modules/input/chinese/packages.el (hunk 3 above)
Comment (merrickluo) [reply modules/input/chinese/packages.el]: pinded to latest commit. I see that multiple module has `(package! posframe :pin "")`, what would happen if they pin different versions?

**Comment by merrickluo** [comment]
Thanks, @hlissner, I updated the PR addressing your comments, in your comments you mentioned #6382 which is the link to this PR, do you mean #6356 ?  
I see they use `pinyinlib` to achieve the same functionality as `pyim-cregexp-build`, so it's not needed, I'll try out the `evil-pinyin` package.

File: modules/input/chinese/packages.el (hunk 3 above)
Comment (hlissner) [reply modules/input/chinese/packages.el]: Then the module that loads last will win. It's not ideal, but it's all we got at the moment.

Comment (merrickluo) [reply modules/input/chinese/packages.el]: I see, thanks for clarifying.

**Comment by merrickluo** [comment]
I tried `evil-pinyin` and it's working great, so I copied the code from #6356 and made it only on with `evil +everywhere`.
Also, I see that the doc rewrite is now done, so I added some documentation about the new flags/packages.
cc @ErnestDong

**Comment by ErnestDong** [comment]
thanks for your hard working

**Comment by hlissner** [comment]
Sorry for the late response. I've merged this in separate parts to make few minor refactors, conformed the commit messages to our git conventions, and resolve some merge conflicts. Let me know if it gives you any trouble. Thanks for your help, in any case!

**Comment by sdvcrx** [comment]
@hlissner  Hi! `doom sync` error if enable `chinese` module after pulling latest Doom Emacs config:

init.el:

```elisp
(doom! :input
       ;;bidi              ; (tfel ot) thgir etirw uoy gnipleh
       chinese
       ;;japanese
       ;;layout            ; auie,ctsrnm is the superior home row
;; ..........
```

First run `doom sync`: 

```
> Synchronizing "default" profile...
  > Regenerating envvars file
    ✓ Generated ~/.emacs.d/.local/env
  > Installing packages...
    > Updating recipe repos...
    > Cloning evil-pinyin... emacsmirror-mirror...
    - Checked out evil-pinyin: 3e9e501ded86f88e01a4edec5d526ab0fab879d7
    > Building evil-pinyin...
    ! Script was abruptly aborted, leaving Doom in an incomplete state!
    - Run 'doom sync' to repair it.
    ✓ Finished in 9.71810s
```

And run `doom sync` again:

```
> Synchronizing "default" profile...
  > Regenerating envvars file
    ✓ Generated ~/.emacs.d/.local/env
  x There was an unexpected runtime error
    Message: Symbol's function definition is void
    Details: (define-namespace)
    Backtrace:
      (define-namespace evil-pinyin- (defvar with-search-rule 'custom "Enabl...
      (load-with-code-conversion "/Users/sdvcrx/.emacs.d/.local/straight/bui...
      (load "/Users/sdvcrx/.emacs.d/.local/straight/build-29.0.60/evil-pinyi...
      (progn (load autoloads-file nil 'nomessage))
      (if (file-exists-p autoloads-file) (progn (load autoloads-file nil 'no...
      (let ((autoloads-file (straight--autoloads-file package))) (if (file-e...
      (straight--load-package-autoloads "evil-pinyin")
      (let ((name (symbol-name package))) (add-to-list 'load-path (directory...
      (let ((package (car tail))) (straight-register-package package) (let (...
      (while tail (let ((package (car tail))) (straight-register-package pac...
      (let ((tail (cl-delete-duplicates packages :test #'equal))) (while tai...
      (let (packages) (let ((tail doom-packages)) (while tail (let ((package...
    ! Wrote extended backtrace to ~/.emacs.d/.local/logs/cli.doom.230223180637.48268.error
  ! Script was abruptly aborted, leaving Doom in an incomplete state!
  - Run 'doom sync' to repair it.
```

My emacs version is 29 and with `native-comp`.

System Information: https://pastebin.com/smwcG0vc

**Comment by merrickluo** [comment]
Thanks @hlissner ! I'm having the same issue as @sdvcrx, seems like `evil-pinyin` can't load `names` when compile. I didn't have this issue previously.

**Comment by hlissner** [comment]
After looking into this, the issue runs deep. Possibly to do with Emacs 29's new loaddefs-gen library (which replaces autoload.el); a new library for generating autoloads files. In any case, since they aren't needed, I've disabled generation of evil-pinyin's autoloads file, which seems to work around this. Give it a try and let me know if that works.

**Comment by merrickluo** [comment]
yeah, it syncs without a problem now,  `evil-pinyin`'s function works great as well. thanks @hlissner :bow:

**Comment by sdvcrx** [comment]
Working perfectly, thanks @hlissner @merrickluo !

**Comment by yangsheng6810** [comment]
The following addition causes problem for users who enable both ivy and Chinese module. The command `pyim-cregexp-ivy` is defined in feature `pyim-cregexp-utils.el`, which [needs to be manually required](https://github.com/tumashu/pyim/blob/f70c4227752dfa53cd3922584624da2484346d53/README.org#%E8%AE%A9-ivy-%E6%94%AF%E6%8C%81%E6%8B%BC%E9%9F%B3%E6%90%9C%E7%B4%A2%E5%80%99%E9%80%89%E9%A1%B9%E5%8A%9F%E8%83%BD). If not required, ivy will try to use `pyim-cregexp-ivy` for all selection commands, which ends in error and makes Emacs almost unusable (try think of doing something without invoking any command that requires selection).

```emacs-lisp
(setq ivy-re-builders-alist
               '((t . pyim-cregexp-ivy)))
```
I currently workaround the issue with the following in my config file:
```emacs-lisp
(use-package pyim-cregexp-utils
  :defer t
  :commands (pyim-cregexp-ivy))
```
Unfortunately I am not familiar with how autoloading works in Doom Emacs, and it would be great if someone could make a fix in Doom itself.

**Pull request closed without merging**